### ISSUE TITLE
use RegExp values in state, but the result is unexpected.

### ISSUE THREAD

**Issue by wendy81** [facebook/react#17129]
a question
**What is the current behavior?**

here is the code.
https://codesandbox.io/s/display-regexp-values-in-propsstate-3wy9o

 ```
import React, { useState, useEffect } from "react";
import ReactDOM from "react-dom";

const useInput = () => {
  const [value, setValue] = useState("");
  const changeEvent = e => {
    setValue(e.target.value);
  };
  const inputCom = <input onChange={changeEvent} value={value} />;
  return [value, inputCom];
};

const MyComp = () => {
  const [isTrue, setIsTrue] = useState(/a/gi);
  const [value, getInput] = useInput();
  const result = isTrue.test(value);

  return (
   <div>
      <div>Q:Does input value include 'abc'?</div>
      <div>A:{result ? "true" : "false"}</div>
      {getInput}
    </div>
  );
};

const rootElement = document.getElementById("root");
ReactDOM.render(<MyComp />, rootElement);
```

 I found if the input value include 1 `a`, 1 true and 1 false will alternating display.
 I found if the input value include 2 `a`, 2 true and 1 false will alternating display.
 I found if the input value include 3 `a`, 3 true and 1 false will alternating display.
......

I'm so confused, I don't why?

**Comment by Sharakai** [comment]
This is because JS RegExp has a `lastIndex` internal state when using the `g` flag. Each call to `test()` uses the `lastIndex` to decide where to start, and then changes that to the index of the next match.

The easiest way to fix this is simply to remove the `g` flag, which'll make this code snippet work as you expect.

You can find more information/discussion on this here: https://stackoverflow.com/questions/1520800/why-does-a-regexp-with-global-flag-give-wrong-results

TL;DR: Not a problem with RegExp in state, but rather a complexity of JS's RegExp's `exec` method.

**Comment by wendy81** [comment]
@Sharakai   thx, I got it.
